Styling: getFadedOverflowStyle function to replace ellipsis

Vitalius1 opened the pull request:
#### Pull request checklist
- [X] Include a change request file using `$ npm run change`

#### Description of changes

Adds a new function returning a style for an :after pseudo element to replace an ellipsis text-overflow with a fading out effect of last characters. Also can be used to apply a vertical fade out for multiline content.
Considers the fact that all colors are affected by the theme applied to the app.

The initial request for this feature came from DetailsList team as part of fluent effort. The intent is not to use ellipsis for the text-overflow but a short 20px fade out gradient. The downside is that it will always be present at the end of the element containing the content even if it's not overflowing. 
- A solution to this would be checking if the container has horizontal scroll, but in case of DetailsList when we could have 1000 items on the page, performance can take a hit. So designers agreed to keep it always present as it will cover at most only the last 3 characters.
- Another solution I tested is creating a wrapper component which can leverage flex-box properties to only place the gradient when we have an overflow. The downside of this approach is that it requires to know a lot of information about the current background state(hover, focus, selected...) which would need to be provided by the developer making it cumbersome to use. Plus it adds 4 div elements to each overflowing item always present in the DOM which again in case of DetailsList use it might be 1000 items. 

#### Focus areas to test

Have tested it with DetailsList examples for horizontal and vertical overflow.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/6115)

phkuo commented on packages/styling/src/styles/getFadedOverflowStyle.ts:
```diff
@@ -0,0 +1,114 @@
+ */
+export function getFadedOverflowStyle(
+  theme: ITheme,
+  color: string = theme.semanticColors ? theme.semanticColors.bodyBackground : DEFAULT_COLOR,
```
i don't think we should have a fallback default color, we want it to fail if the theme is incorrect, since it's an indication of bigger problems

phkuo commented on packages/styling/src/styles/getFadedOverflowStyle.ts:
```diff
@@ -0,0 +1,114 @@
+ *
+ * @param color - Color to be converted from hex to rgba.
+ */
+function hex2rgb(color: string): IRGB {
```
all of these should sit in or next to colors.ts
note that colors.ts is quite large since it pulls in all the named colors... so we might want to make a separate minimal color utilities file that doesn't include that

Vitalius1 replied:
@phkuo colors.ts is a part of office-ui-fabric-react package which is not a dependency of styling package where my function resides. This was the main reason I've added this helper function. Should we move the entire colors folder from OUFR to Styling and leave just the index importing from it to prevent a breaking change?

phkuo commented on packages/styling/src/styles/getFadedOverflowStyle.ts:
```diff
@@ -0,0 +1,113 @@
+ */
+export function getFadedOverflowStyle(
+  theme: ITheme,
+  color: string = theme.semanticColors.bodyBackground,
```
this should be something in typeof ISemanticColors (whatever the syntax is for that)

Vitalius1 replied:
@phkuo user is passing only the value of a semantic color which is a string... don't understand what do you mean by typeof ISemanticColors. Am I missing something?

phkuo replied:
@Vitalius1 we have other places (see the text thing in experiments) which enforce that you pass in a semantic slot, so you're actually passing in a type. And then it will get that slot from the theme for you. It's a way to force users to use only theme colors, instead of custom ones. You can consider doing that here, instead of allowing raw colors. Up to you.

Vitalius1 commented:
@phkuo I changed the type on color parameter as requested. Also added some more logic to the color conversion helper function. We really need to create color utility module in Styling package.
Would appreciate if could take another pass to review the latest changes.

phkuo commented:
@Vitalius1 looks great, thanks!
